### PR TITLE
Move DeviceLimits validation checks into Core/Parameter_Validation layers

### PR DESCRIPTION
**Pull request by mark-lunarg**
Move the stateful checks into CV and the stateless-ish checks into PV.  Note that this change leaves the DeviceLimits layer intact as it will be removed in a subsequent PR.

### CODE REVIEW

File: layers/core_validation.cpp [hunk 1]
```diff
@@ -145,10 +149,13 @@ struct layer_data {
...
     PHYS_DEV_PROPERTIES_NODE phys_dev_properties;
     VkPhysicalDeviceMemoryProperties phys_dev_mem_props;
+    VkPhysicalDeviceFeatures physical_device_features;
+    unique_ptr<PHYSICAL_DEVICE_STATE> physical_device_state;
 
     layer_data()
-        : report_data(nullptr), device_dispatch_table(nullptr), instance_dispatch_table(nullptr), device_extensions(),
-          device(VK_NULL_HANDLE), phys_dev_properties{}, phys_dev_mem_props{} {};
+        : instance_state(nullptr), report_data(nullptr), device_dispatch_table(nullptr), instance_dispatch_table(nullptr),
+          device_extensions(), device(VK_NULL_HANDLE), phys_dev_properties{}, phys_dev_mem_props{},
```
Comment (tobine) [inline layers/core_validation.cpp]: Add "physical_device_features{}" to init list as well?

File: layers/core_validation.cpp [hunk 2]
```diff
@@ -3939,7 +3946,7 @@ CreateInstance(const VkInstanceCreateInfo *pCreateInfo, const VkAllocationCallba
                                      pCreateInfo->ppEnabledExtensionNames);
 
     init_core_validation(instance_data, pAllocator);
-
+    instance_data->instance_state = unique_ptr<INSTANCE_STATE>(new INSTANCE_STATE());
```
Comment (dustin-lunarg) [inline layers/core_validation.cpp]: You can use std::make_unique here:
 instance_data->instance_state = std::make_unique<INSTANCE_STATE>();

File: layers/core_validation.cpp [hunk 3]
```diff
@@ -3989,9 +3996,88 @@ static void createDeviceRegisterExtensions(const VkDeviceCreateInfo *pCreateInfo
...
+        }
+    }
+    return skip_call;
+}
+
+// Verify that features have been queried and that they are available
+static bool ValidateRequestedFeatures(layer_data *dev_data, const VkPhysicalDeviceFeatures *requested_features) {
+    bool skip_call = false;
+
+    VkBool32 *actual = (VkBool32 *)&(dev_data->physical_device_features);
```
Comment (dustin-lunarg) [inline layers/core_validation.cpp]: reinterpret_cast

**Comment by tobine** [comment]
High level ?. You say you're moving the checks, but I don't see that device_limits.cpp/h are actually being removed. Is that coming in a separate commit or am I missing it? Also is there a corresponding update for desktop loader to modify LunarG standard_validation layer?

**Comment by mark-lunarg** [comment]
Tobin -- Yeah, I mentioned it in the PR description.  This is independent of removing that layer and the changes co-exist right now.  That'll be a separate patch series to follow.

File: layers/parameter_validation.cpp [hunk 4]
```diff
@@ -3498,6 +3515,38 @@ UpdateDescriptorSets(VkDevice device, uint32_t descriptorWriteCount, const VkWri
...
+
+            if ((pDescriptorWrites[i].descriptorType == VK_DESCRIPTOR_TYPE_UNIFORM_BUFFER) ||
+                (pDescriptorWrites[i].descriptorType == VK_DESCRIPTOR_TYPE_UNIFORM_BUFFER_DYNAMIC)) {
+                VkDeviceSize uniformAlignment = device_data->device_limits.minUniformBufferOffsetAlignment;
+                for (uint32_t j = 0; j < pDescriptorWrites[i].descriptorCount; j++) {
+                    if (pDescriptorWrites[i].pBufferInfo != NULL) {
+                        if (vk_safe_modulo(pDescriptorWrites[i].pBufferInfo[j].offset, uniformAlignment) != 0) {
+                            skip_call |=
+                                log_msg(device_data->report_data, VK_DEBUG_REPORT_ERROR_BIT_EXT,
+                                        VK_DEBUG_REPORT_OBJECT_TYPE_PHYSICAL_DEVICE_EXT, 0, __LINE__, DEVICE_LIMIT, "PARAMCHECK",
```
Comment (dustin-lunarg) [inline layers/parameter_validation.cpp]: Use the LayerName constant instead of the "PARAMCHECK" literal

File: layers/core_validation.cpp (hunk 2 above)
Comment (mark-lunarg) [inline layers/core_validation.cpp]: That's cool!  There is another spot where I was able to do the same.

File: layers/core_validation.cpp (hunk 3 above)
Comment (mark-lunarg) [inline layers/core_validation.cpp]: Got them both.

File: layers/parameter_validation.cpp (hunk 4 above)
Comment (mark-lunarg) [inline layers/parameter_validation.cpp]: Went ahead and did all of them in the layer.

File: layers/core_validation.cpp (hunk 1 above)
Comment (mark-lunarg) [inline layers/core_validation.cpp]: Done.

File: layers/vk_validation_layer_details.md [hunk 5]
```diff
@@ -185,6 +198,8 @@ The VK_LAYER_LUNARG_parameter_validation layer validates parameter values and fl
 | Required Parameter | Verifies that a required parameter was not specified as 0 or NULL | REQUIRED_PARAMETER | | RequiredParameter | NA |
 | Reserved Parameter | Verifies that a parameter reserved for future use was specified as 0 or NULL | RESERVED_PARAMETER | | ReservedParameter | NA |
 | Unrecognized Value | Verifies that a Vulkan enumeration, VkFlags, or VkBool32 parameter contains a value that is recognized as valid for that type | UNRECOGNIZED_VALUE | | UnrecognizedValue | NA |
+| Device Limit Violation | Verifies that a parameter is within the limits advertised by the gpu | DEVICE_LIMIT | vkUpdateDescriptorSets vkCreateRenderPass | Todo | NA |
+| Device Feature Violation | Verifies that a requested feature is supported by the gpu | DEVICE_FEATURE | vkBeginCommandBuffer | Todo | NA |
 | Failed Call Return Code | Provides a description of a failure code returned by a Vulkan API call | FAILURE_RETURN_CODE | | FailedReturnValue | NA |
```
Comment (tobine) [inline layers/vk_validation_layer_details.md]: On Windows, the "TODO" check is complaining about "Todo". Can you update these to uppercase "TODO"?

Comment (mark-lunarg) [inline layers/vk_validation_layer_details.md]: I had already done this on my local branch.

File: layers/core_validation.cpp (hunk 2 above)
Comment (mark-lunarg) [inline layers/core_validation.cpp]: Ah, bummer, this is c++14 only -- linux and android barf.  Switching back.

**Comment by mark-lunarg** [comment]
Cherry-picked to Master.

**Pull request closed without merging**